README.md
# SAFDOM
Simple as $uck DOM API shorthands. Fast as hell. Mainly to use with latest Chromium (24+).
All other browsers compatibility notes in code are JUST FOR REFERENCE, and not all features have such notes.
It is not intended to be totally cross-browser, although tries to use only general, well-known and compatible calls.

**API IS ABOUT TO CHANGE in future releases, so use at your own risk, or use fixed version**

I'm using it with my [nw.js](http://nwjs.io/) projects to get totally rid of jQuery (which is VERY SLOW and bloated for such use),
while having convinient tool for querying CSS selectors, DOM traversal and manipulations.

It does not have tests (yet?) and had been published only for demonstration purposes.
Can have some dirty comments and untested code and does not ready for production use.

About tests: having the code not being very complicated, it still requires some unit testing.
If you can/want to help writing them (and for any other contribution or sugestions) -- use offshore@aopdg.com to contact me.

P.S. I know that extending core objects' prototypes is far not a 'best practice' and have some side effects,
but hey. Accept it or leave it. It's a specific thing for a specific case.
And bear in mind that unlike other browsers, Chromium is immune to performance impact introduced by such approach.

## Dive in

Just
```
npm install safdom
```
or
```
git clone https://github.com/offshore/safdom.git
```

SAFDOM default behavior depends on include method.
The detection based on variable `module` tested to be in surrounding context.
See below.

### Compatibility
That's a shallow compatibility table for major features; some of browsers' lower versions are not mentioned.

| Browser   | Version       | Feature           | Dependency
| --------- | ------------- | ----------------- | ----------
| Chrome    | **24+**       | `.$class*` full   | [`Element.classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList): several arguments in the add/remove/toggle
| Chrome    | 8+            | `.$class*` basic  | [`Element.classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList)
| FF        | **26+**       | `.$class*` full   | [`Element.classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList): several arguments in the add/remove/toggle
| FF        | 13+           | `.$clone(true)`   | [`Node.cloneNode(true)`](https://developer.mozilla.org/en-US/docs/Web/API/Node/cloneNode)
| FF        | 9+            | `.$has`           | [`Node.contains`](https://developer.mozilla.org/en-US/docs/Web/API/Node/contains)
| FF        | 3.6+          | `.$class*` basic  | [`Element.classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList)
| IE        | **10+**       | `.$class*` full   | [`Element.classList`](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList): several arguments in the add/remove/toggle
| IE        | 9+            |                   | Too many things became available only with IE 9+
| Opera     | 11.5+, **15.0+** | `.$is`         | [`Element.matches`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches)
| Safari    | **5.0+**      | `.$is`            | [`Element.matches`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches)

### Browsers / simple
```html
<script type="text/javascript" src="path/to/dist/safdom.min.js"></script>
```
or CDN:
```html
<script type="text/javascript" src="//npmcdn.com/safdom@0.2.3/dist/safdom.min.js"></script>
```

This just loads SAFDOM and do prototype extension automatically.

### NW.js / advanced
If your nw.js project is simple single-window application, you can use the browser method already described above.
But there is one more interesting option available, especially if you open (possibly many) windows dynamically.
In background context you can do something like this:
```javascript
var $SAF = require('safdom');
// when you open new window, do magic:
nw.Window.open('someWindow.html', {}, function(someWindow) {
    $SAF(someWindow.window);
});
```
And then, `someWindow`'s DOM objects got extended.

## API Reference

### A note on forEach, NodeList and HTMLCollection
The only unprefixed thing in SAFDOM is a `forEach` method for NodeList and HTMLCollection;
however its done to achieve the same behavior and interface as seen on `(Array|Map|Set|...).forEach()`;
the only difference is that it returns `list` on which it was called -- for chaining purposes.
- **`list.forEach(callback, thisArg)`**: wrapper to `Array.prototype.forEach.call(list, callback, thisArg || list)`, returns list;
  `callback`'s `this` is pointing to `thisArg`; and arguments are:
  - `v`: (currentValue), the current thing being processed in the list
  - `k`: (index), the index of the current element being processed in the list
  - `list`: (array), The array that `forEach` is being applied to

Another method for NodeList and HTMLCollection is `$slice`, which acts exactly as `Array.slice`
and can be used to flatten LIVE lists.
- **`list.$slice(begin, end)`**: same as `Array.prototype.slice.call(list, begin, end)`, returns shallow copied list transformed to plain array

Maybe it's worth switching from `forEach` to `map`, because latter is more suitable for chaining.
Anyway, I think NodeList and HTMLCollection deserve their own `map` and `reduce` methods.

### Node comparison
- **`x.$has(y)`**: same as [`x.contains(y)`](https://developer.mozilla.org/en-US/docs/Web/API/Node/contains)
- **`x.$eq(y)`**: check `x` identity (===) to `y`
- **`x.$lt(y)`** and **`x.$gt(y)`**: determine that `x` is somewhere before or after `y` in DOM tree, respectively.
  Uses [`Node.compareDocumentPosition`](https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition), returns `0` or `non-0`
  (see `Node.DOCUMENT_POSITION_*`).

### Node manipulation
All of those methods do the same things as their jQuery cousins;
except that `x` and `y` can only be `Node` instances (no strings, selectors, arrays or other magic handling).
However you may be interested in `$(html|text)(Before|Prepend|Append|After)` -- see below.
They all return `x` (except `$clone`, which returns (optionally deep) detached copy of `x`), so can be easily chained with each other or other manipulation shorthands.
- **`x.$after(y)`** \* x must have parent
- **`x.$append(y)`**
- **`x.$appendTo(y)`**
- **`x.$before(y)`** \* x must have parent
- **`x.$clone(deep)`**: same as [`x.cloneNode(deep)`](https://developer.mozilla.org/en-US/docs/Web/API/Node/cloneNode); returns detached copy of x
- **`x.$detach()`**
- **`x.$insertAfter(y)`** \* y must have parent
- **`x.$insertBefore(y)`** \* y  must have parent
- **`x.$prepend(y)`**
- **`x.$prependTo(y)`**
- **`x.$replace(y)`** \* x must have parent; remember that [`replaceChild`](https://developer.mozilla.org/en-US/docs/Web/API/Node/replaceChild) is destructive
- **`x.$clear()`**: same as jQuery.empty() -- wipes children

### Element extensions
#### Attributes
- **`x.$attrH(k)`**: same as [`x.hasAttribute(k)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/hasAttribute)
- **`x.$attrG(k)`**: same as [`x.getAttribute(k)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getAttribute), pay attention to behavior notes in MDN related to return value and lowercasing
- **`x.$attrS(k, v)`**: same as [`x.setAttribute(k, v)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute), but returns x for chaining
- **`x.$attrD(k, k, ...)`**: same as [`x.removeAttribute(k)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/removeAttribute) for each argument, but returns x for chaining
- **`x.$attrU(k, v)`**: $attrD when v == null, $attrS otherwise
- **`x.$attrU(kv)`**: $attrU for each k and v in kv

#### Properties
- **`x.$propG(k)`**: just return x[k]
- **`x.$propS(k, v)`**: set x[k] to v, return x for chaining

#### Class names
Remember that IE10- does not support the second parameter for `.toggle()` and multiple params for `.add` or `.remove`;
however it may be shimmed later. See docs for [Element.classList](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) / [DOMTokenList](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList).
- **`x.$classA(v, v, ...)`**: add each v to `Element.classList`
- **`x.$classD(v, v, ...)`**: remove each v from `Element.classList`
- **`x.$classH(v)`**: check that `Element.classList` contains v
- **`x.$classT(v, state)`**: toggle v in `Element.classList` when state == null, otherwise add/remove it for state is true/false, respectively

#### HTML content and manipulation
- **`x.$htmlG()`**: just return x.innerHTML
- **`x.$htmlS(v)`**: set x.innerHTML to v, return x for chaining

Those are crafty. Allows you to update HTML content inside and around element without side effects (losing events or performance impact);
so they act like `x.$before`, `x.$prepend`, `x.$append` and `x.$after` versions with html argument.
See [`Element.insertAdjacentHTML`](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentHTML) for details.
- **`x.$htmlBefore(v)`** \* x must have parent
- **`x.$htmlPrepend(v)`**: like `x.innerHTML = v + x.innerHTML`
- **`x.$htmlAppend(v)`**: like `x.innerHTML = x.innerHTML + v`
- **`x.$htmlAfter(v)`** \* x must have parent

#### Text content
Note compatibility issues for \*G and \*S: see [reference for `Node.textContent`](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent).
- **`x.$textG(k)`**: just return x.textContent
- **`x.$textS(k, v)`**: set x.textContent to v, return x for chaining

Acts like `x.$before`, `x.$prepend`, `x.$append` and `x.$after` versions with text argument, automatically creating text Node with supplied
argument and attaching it appropriately.
- **`x.$textBefore(v)`**: \* x must have parent
- **`x.$textPrepend(v)`**
- **`x.$textAppend(v)`**
- **`x.$textAfter(v)`**: \* x must have parent

#### Search/selection utilities
Note that results of methods for querying multiple things can return not only HTMLCollection, but NodeList as well, and both LIVE and NON-live versions;
it can be tricky sometimes, so be sure to check documented behavior for appropriate shorthands.
- **`x.$f(selector)`**: same as [`x.querySelectorAll(selector)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelectorAll), returns NON-live list
- **`x.$s(selector)`**: same as [`x.querySelector(selector)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector), returns first match or `null`
- **`x.$t(tagName)`**: same as [`x.getElementsByTagName(tagName)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getElementsByTagName), returns LIVE list
- **`x.$c(className)`**: same as [`x.getElementsByClassName(className)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getElementsByClassName), returns LIVE list
- **`x.$is(selector)`**: same as [`x.matches(selector)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches), returns bool
- **`x.$closest(selector)`**: same as [`x.matches(selector)`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches), returns closest matching ancestor (including x) or null

#### Traversal utils related to Element (not Node!)
- **`x.$idx()`**: obtain x index relative to its parent, or -1 if not found or detached
- **`x.$idxOf(y)`**: obtain y index relative to x, which should be it's parent, returns -1 if not found, detached, or y is not direct child of x
- **`x.$up()`**: just return `x.parentElement`
- **`x.$fc()`**: just return `x.firstElementClild`
- **`x.$lc()`**: just return `x.lastElementClild`
- **`x.$ps()`**: just return `x.previousElementSibling`
- **`x.$ns()`**: just return `x.nextElementSibling`

**Deprecated for now** due to naming problem and tricky nature (uses css4 `:scope` and index detection, not mentioned in compatibility table).
Most likely be renamed to `$psa` and `$nsa`, or even be completely cut:
- **`x.$pss()`**: find all siblings between (and including if not null) `x.parentElement.firstElementChild` and `x.previousElementSibling`, returns NONlive list
- **`x.$nss()`**: find all siblings between (and including if not null) `x.nextElementSibling` and `x.parentElement.lastElementChild`, returns NONlive list

#### More utils
- **`x.$cs(pseudoElt)`**: same as [`window.getComputedStyle(x, pseudoElt)`](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle)
- **`x.$rect()`**: same as [`x.getBoundingClientRect()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect)
- **`x.$rects()`**: same as [`x.getClientRects()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getClientRects)

See also `$meta.px()`/`$meta.py()`.

### EventTarget shorthands
**Only minimal set of utilities for now**
Note that EventTarget.prototype is not consistently available, so fo browsers like IE, SAFDOM extends only `window`, `window.document` and `Node.prototype` with related methods,
leaving other thing like XMLHttpRequest unmodified.
- **`x.$on(event, handler)`**: same as [`x.addEventListener(event, handler, false)`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener), but returns x for chaining
- **`x.$onc(event, handler)`**: capturing version for `$on`
- **`x.$off(event, handler)`**: same as [`x.removeEventListener(event, handler, false)`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener), but returns x for chaining
- **`x.$offc(event, handler)`**: capturing version for `$off`

### Globals
**Selectors**:
- **`$id(id)`**: same as [`document.getElementById(id)`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementById), returns thing or `null`
- **`$f(selector)`**: same as [`document.querySelectorAll(selector)`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll), returns NON-live list
- **`$s(selector)`**: same as [`document.querySelector(selector)`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector), returns first match or `null`
- **`$t(tagName)`**: same as [`document.getElementsByTagName(tagName)`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByTagName), returns LIVE list
- **`$c(className)`**: same as [`document.getElementsByClassName(className)`](https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByClassName), returns LIVE list

**Minimal (for now) construction shorthands**:
- **`$mk(tagName)`**: same as [`document.createElement(tagName)`](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement)
- **`$mkT(textContent)`**: same as [`document.createTextNode(textContent)`](https://developer.mozilla.org/en-US/docs/Web/API/Document/createTextNode)
- **`$mkF()`**: same as [`document.createDocumentFragment()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/createDocumentFragment)

**Utility**:

It really hurts when you try to determine, which element is actually the main scrolling thing, `html` or `body`.
Here, take this painkiller.
- **`$meta.se()`**: utility to get [`document.scrollingElement`](https://developer.mozilla.org/en-US/docs/Web/API/Document/scrollingElement). Includes cross-browser shim via https://github.com/mathiasbynens/document.scrollingElement, which is partially rewritten and optimized to be fully transparent in Chromium. This method caches it's return value, so the steps to detect scrollingElement are not performed once `document.readyState` is no more `loading`. I guess it's useful for weird setups only, but if you need to flush this cache, call `$meta.seReread()`.
- **`$meta.px()`, `$meta.py()`**: viewport pageXOffset/pageXOffset to use with `Element.$rect`/`Element.$rects` (in general, this is not the same as scrollLeft/scrollTop, but workaround included; see also https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollY). Note that this functions are affected by scrollingElement cache (see `$meta.se()`).


### Appendix: method name parts cheatsheet
The dictionary of method names is fairly simple, short and easy to remember; screw those who prefer `WaitForMultipleObjects` over `poll`.
- attr: **Attr**ibute-related stuff
- class: Element **class** names
- html: **HTML** content manipulation
- prop: **Prop**erties

CHANGELOG.md
# SAFDOM changelog

## 0.2.3 (2016-07-12)
* Considered unuseful in $meta and has been cut: sl*, st*, vw, vh, cw, ch, sw, sh. Use respective `scrollLeft`, `scrollHeight`, etc on `$meta.se()`.
* The scrollingElement accessor `$meta.se()` now caches it's return value. Use `$meta.seReread()` to flush it (I guess it's useful for weird setups only).

## 0.2.2 (2016-07-01)
* Included $rect and $rects, which are aliases for Element.getBoundingClientRect and Element.getClientRects, respectively.
* Included minimalistic getComputedStyle utility as Element.$cs
* New global: $meta, which I suppose to use for utility methods such as browser/window/context/viewport/document utils/stats detection/manipulation. Kinda messy.
* Methods for main viewport inspection in $meta (client/inner/scroll-related dimensions)
* $meta.se() to polyfill document.scrollingElement via https://github.com/mathiasbynens/document.scrollingElement
* So, the as noted in library description, it is still optimized for Chromium; I realize that wider browser support is needed, so in future some shims will be introduced.
* Therefore, SAFDOM API is about to change: at least, all globals most likely will be migrated into one global `$SAF` object and something like jQuery.noConflict provided to handle special cases.
* Also planned: utils/aliases (NOT POLYFILLS OR SHIMS) for custom Elements, HTML Templates, Shadow DOM and HTML Imports.

## 0.2.1 (2016-03-09)
* Included fallback for Element.contains, minimal Chrome version is now 24+
* Included fallback for Element.matches
* $mkF for document.createDocumentFragment
* Documentation fixes and compatibility table

## 0.2.0 (2016-03-05)
* Different behavior for various environments
* Documentation fixes

## 0.1.0 (2016-03-04)
* It was decided to use `forEach` instead of `$each` for NodeList and HTMLCollection
* It was decided to remove console.$time and console.$timeEnd
* Near full usage reference provided
* Introducing changelog

## 0.0.1 (2016-03-03)
* Initial revision


src/safdom.js
/*! SAFDOM | (c) Alex Offshore | aopdg.com */
;(function(module, exports) {

/**
* PRIMARY TARGET IS LATEST CHROMIUM (Chrome 45+), shorthands and calls optimized for use with it
* all other browsers compatibility notes are JUST FOR REFERENCE, and not for all features,
* although latest versions of FF, Opera, Safari and even IE (Edge) can handle most things fine.
* Feature       Chrome min version  Notes
* $f            1+
* $1            1+
* $t            1+
* $is           34+
* $closest      45+
* $idx, $idxOf  1+                  Dependent on ParentNode.children
* $fc, $lc      1+
* $ps, $ns      4+
* $pss          20+
* $nss          20+
* $class*       8+                  Dependent on Element.classList
*/

// `Object.assign` is not supported by IE w/o polyfill (but Edge has it), others: [Chrome 45+, FF 34+, Opera 32+, Safari 9+]
// however, the following IS NOT A POLYFILL, just a shorthand
function _extendX(o) {
    for (var k, ai = 1; a = arguments[ai]; ++ai) for (k in a) if (a.hasOwnProperty(k)) o[k] = a[k];
    return o;
};
function _extend(ctx) { return ctx.Object.assign || _extendX; };

// internal utility methods:
function _util_compatMode(ctx) {
    return (ctx.document.compatMode || '') === 'CSS1Compat';
};
// scrollElement shim via https://github.com/mathiasbynens/document.scrollingElement, partially rewritten
// see also: http://www.quirksmode.org/mobile/tableViewport.html
function _util_mkScrollingElementGetter(ctx) {
    var document = ctx.document;
    if ('scrollingElement' in document) return true; // woohoo, simple case, no dedicated getter needed
    // Well, we need a runtime detector for that. So frustrating.
    var isBodyElement = ctx.HTMLBodyElement
        ? function(elt) { return elt instanceof HTMLBodyElement; }
        : function(elt) { return elt.tagName.toLocaleLowerCase() == 'body'; }; // `tagName` is uppercase in HTML, but may be any-case in XML
    var locateBodyElement = function(frameset) {
        // We use this function to be correct per spec in case `document.body` is
        // a `frameset` but there exists a later `body`. Since `document.body` is
        // a `frameset`, we know the root is an `html`, and there was no `body`
        // before the `frameset`, so we just need to look at siblings after the
        // `frameset`.
        var current = frameset;
        do {
            if (current.nodeType == 1 && isBodyElement(current)) return current;
        } while (current = current.nextSibling);
        return null; // nothing found
    };
    var isScrollingElementCompliant = function() {
        // In quirks mode, the result is equivalent to the non-compliant standards mode behavior.
        if (!_util_compatMode(ctx)) return false;
        var iframe = document.createElement('iframe');
        iframe.style.height = '1px';
        (document.body || document.documentElement || document).appendChild(iframe);
        var iframeDoc = iframe.contentWindow.document;
        iframeDoc.write('<!DOCTYPE html><div style="height:9999em">x</div>');
        iframeDoc.close();
        var result = iframeDoc.documentElement.scrollHeight > iframeDoc.body.scrollHeight;
        iframe.parentNode.removeChild(iframe);
        isScrollingElementCompliant = function() { return _util_compatMode(ctx) && result; }; // further calls faster
        return result;
    };
    var isScrollable = function(body) {
        // a `body` element is scrollable if `body` and `html` both have non-`visible` overflow and are both being rendered
        // about check that element is rendered:
        // I've cut the visibility == 'collapse' test for table-*, it can be things-breaking for some setups
        // or maybe do I check offsetHeight and offsetParent values instead messing with styles?
        //TODO investigate that
        var bodyStyle = elt.$cs(),
            htmlStyle = document.documentElement.$cs();
        return bodyStyle.overflow != 'visible' && bodyStyle.display != 'none' && bodyStyle.visibility == 'visible'
            && htmlStyle.overflow != 'visible' && htmlStyle.display != 'none' && htmlStyle.visibility == 'visible';
    };
    return function() {
        if (isScrollingElementCompliant()) return document.documentElement; // better than it can be
        // Need duck typing body element and check other heavy shit. Every fucking time. R.I.P. performance.
        // you should note that document.body could be not only a HTMLBodyElement,
        // but at least HTMLFrameSetElement or null (e.g. if not yet initialized)
        var body = document.body && locateBodyElement(document.body);
        // If `body` is itself scrollable, it is not the `scrollingElement`.
        return body && isScrollable(body) ? null : body;
    };
};

// prototype extensions:
function _protoNode(ctx) {
    var Node = ctx.Node,
        DPP = Node.DOCUMENT_POSITION_PRECEDING,
        DPF = Node.DOCUMENT_POSITION_FOLLOWING;

    return {
    // node position comparison stuff
    $has: Node.prototype.contains, // THIS is one of parents of ELT => bool [FF 9+, IE 5+, Safari 3+], https://developer.mozilla.org/en-US/docs/Web/API/Node/contains
    $eq: function(elt) { return this === elt; }, // THIS equals ELT => bool
    $gt: function(elt) { return this.compareDocumentPosition(elt) & DPP; }, // THIS is somewhere after ELT => int [IE 9+]
    $lt: function(elt) { return this.compareDocumentPosition(elt) & DPF; }, // THIS is somewhere before ELT => int [IE 9+]

    // node manipulation, THIS is what node op is performing on, ELT is reference node, returning THIS if not otherwise noted
    $after: function(elt) { return this.parentNode.insertBefore(elt, elt.nextSibling), this; }, // ELT after THIS (this must have parent)
    $append: function(elt) { return this.appendChild(elt), this; }, // ELT as last child of THIS
    $appendTo: function(elt) { return elt.appendChild(this); }, // THIS as last child of ELT
    $before: function(elt) { return this.parentNode.insertBefore(elt, this), this; }, // ELT before THIS (this must have parent)
    $clone: Node.prototype.cloneNode, // clone THIS (optionally deep), return NEW
    $detach: function() { return this.parentNode ? this.parentNode.removeChild(this) : this; }, // detach/remove THIS
    $insertAfter: function(elt) { return elt.parentNode.insertBefore(this, elt.nextSibling); }, // THIS after ELT (elt must have parent)
    $insertBefore: function(elt) { return elt.parentNode.insertBefore(this, elt); }, // THIS after ELT (elt must have parent)
    $prepend: function(elt) { return this.insertBefore(elt, this.firstChild), this; }, // ELT as first child of THIS
    $prependTo: function(elt) { return elt.insertBefore(this, elt.firstChild); }, // THIS as first child of ELT
    $replace: function(elt) { return this.parentNode.replaceChild(elt, this); }, // replace THIS with ELT
    $clear: function() { while (this.lastChild) this.removeChild(this.lastChild); return this; }, // wipe children
}; };

function _protoArr(ctx) {
    var forEach = ctx.Array.prototype.forEach;

    return {
    // enumeration / collection related utils
    forEach: function(callback, thisArg) { return forEach.call(this, callback, thisArg || this), this; }, // callback(element, index, collection)
    $slice: ctx.Array.prototype.slice, // to transform (live) collections to arrays
}; };

function _protoElement(ctx) {
    var protoElement = ctx.Element.prototype,
        dtlA = ctx.DOMTokenList.prototype.add,
        dtlD = ctx.DOMTokenList.prototype.remove,
        forEach = ctx.Array.prototype.forEach,
        indexOf = ctx.Array.prototype.indexOf;

    var matches = protoElement.matches || // is matching selector => bool [Chrome 34+, FF 34+, IE ?, Opera 21+, Safari 7.1+], https://developer.mozilla.org/en-US/docs/Web/API/Element/matches
        protoElement.webkitMatchesSelector || // [Chrome All, Opera 15+]
        protoElement.mozMatchesSelector || // [FF 3.6+]
        protoElement.msMatchesSelector || // [IE 9.0+]
        protoElement.oMatchesSelector || // [Opera 11.5+]
        protoElement.matchesSelector; // general fallback, if available
    var closest = protoElement.closest || // closest ancestor of THIS (or THIS itself) which matches selector => Element or null; native: [Chrome 41+, Firefox 35+], https://developer.mozilla.org/en-US/docs/Web/API/Element/closest
        function(selector) { // polyfill depends on .$is / Element.matches
            var cur = this;
            do {
                if (cur.$is(selector)) return cur;
            } while (cur = cur.parentElement);
            return null;
        };

    return {
    // attributes
    $attrH: protoElement.hasAttribute,
    $attrG: protoElement.getAttribute,
    $attrS: function(k, v) { return this.setAttribute(k, v), this; },
    $attrD: function(k) { return forEach.call(arguments, protoElement.removeAttribute, this), this; }, // supports many arguments
    $attrU: function(kv, v) { // unset when v == null (i.e. null or undefined)
        if (arguments.length == 2) return ((v == null) ? this.removeAttribute(kv) : this.setAttribute(kv, v)), this;
        for (var k in kv) if (kv.hasOwnProperty(k)) ((v = kv[k]), v == null) ? this.removeAttribute(k) : this.setAttribute(k, v);
        return this;
    },

    // class [Chrome 24+, IE10+, FF 24+: Lower versions have no support the second parameter for .toggle() and multiple params for .add / .remove, or completely does not support .classList]
    $classA: function() { return dtlA.apply(this.classList, arguments), this; }, // same as $classAM
    $classD: function() { return dtlD.apply(this.classList, arguments), this; }, // same as $classDM
    $classH: function(token) { return this.classList.contains(token); },
    $classT: function(token, force) { return this.classList.toggle(token, force), this; }, // toggle when force == null (i.e. null or undefined)

    // html content
    $htmlG: function() { return this.innerHTML; },
    $htmlS: function(html) { return (this.innerHTML = html), this; },
    $htmlBefore: function(html) { return this.insertAdjacentHTML('beforebegin', html), this; }, // HTML before THIS (this must have parent)
    $htmlPrepend: function(html) { return this.insertAdjacentHTML('afterbegin', html), this; }, // HTML before first child of THIS
    $htmlAppend: function(html) { return this.insertAdjacentHTML('beforeend', html), this; }, // HTML after last child of THIS
    $htmlAfter: function(html) { return this.insertAdjacentHTML('afterend', html), this; }, // HTML after THIS (this must have parent)

    // properties
    $propG: function(k) { return this[k]; },
    $propS: function(k, v) { return (this[k] = v), this; },

    // text content
    $textG: function() { return this.textContent; },
    $textS: function(text) { return (this.textContent = text), this; },
    $textBefore: function(text) { return this.parentNode.insertBefore(document.createTextNode(text), this), this; }, // TEXT before THIS (this must have parent)
    $textPrepend: function(text) { return this.insertBefore(document.createTextNode(text), this.firstChild), this; }, // TEXT before first child of THIS
    $textAppend: function(text) { return this.appendChild(document.createTextNode(text)), this; }, // TEXT after last child of THIS
    $textAfter: function(text) { return this.parentNode.insertBefore(document.createTextNode(text), this.nextSibling), this; }, // TEXT after THIS (this must have parent)

    // search/selection utils
    $f: protoElement.querySelectorAll, // Find all => NONlive HTMLCollection [Chrome 1+, FF 3.5+, IE 8+, Opera 10+, Safari 3.2+]
    $s: protoElement.querySelector, // Single (first) matched element => Element or null [Chrome 1+, FF 3.5+, IE 8+, Opera 10+, Safari 3.2+]
    $t: protoElement.getElementsByTagName, // by Tag name => LIVE HTMLCollection [Chrome 1+, IE 6+]
    $c: protoElement.getElementsByClassName, // by Class name => LIVE HTMLCollection [IE 9+]
    $is: matches,
    $closest: closest,

    // traversal utils related to Element (not Node)
    $idx: function() { return indexOf.call(this.parentElement.children || [], this); }, // THIS element index relative to its parent, or -1 if not found or detached
    $idxOf: function(elt) { return indexOf.call(this.children, elt); }, // ELT index relative to THIS, or -1 if not a direct child
    $up: function() { return this.parentElement; }, // shorthand to THIS.parentElement
    $fc: function() { return this.firstElementClild; }, // shorthand to THIS.firstElementClild
    $lc: function() { return this.lastElementClild; }, // shorthand to THIS.lastElementClild
    $ps: function() { return this.previousElementSibling; }, // one immidiately previous Element sibling => Element or null
    $ns: function() { return this.nextElementSibling; }, // one immidiately next Element sibling => Element or null

    //TODO rename to $psa/$nsa or completely drop (too complicated)
    $pss: function() { return this.parentElement.querySelectorAll(':scope > :not(:nth-child(' + this.$idx() + ') ~ *)'); }, // all siblings from parentElement.firstElementChild to this.previousElementSibling => NONlive HTMLCollection
    $nss: function() { return this.querySelectorAll(':scope ~ *'); }, // all siblings from this.nextElementSibling, to parentElement.lastElementChild => NONlive HTMLCollection

    // utility
    // alias for getComputedStyle with the minimal "shim" SUDDENLY for IE < 9
    $cs: ctx.getComputedStyle
        ? function(pseudoElt) { return ctx.getComputedStyle(this, pseudoElt || null); }
        : function(pseudoElt) { return this.currentStyle; },

    // aliases for getBoundingClientRect/getClientRects
    // take a look at https://msdn.microsoft.com/en-us/library/hh781509.aspx
    $rect: protoElement.getBoundingClientRect,
    $rects: protoElement.getClientRects,
}; };

function _protoEventTarget(ctx) {
    return {
    // those are shorthands for objects that implement EventTarget interface
    // however, EventTarget.prototype itself may not be available, see $SAF()
    //TODO
    $on: function(event, handler) { return this.addEventListener(event, handler, false), this; },
    $onc: function(event, handler) { return this.addEventListener(event, handler, true), this; },
    $off: function(event, handler) { return this.removeEventListener(event, handler, false), this; },
    $offc: function(event, handler) { return this.removeEventListener(event, handler, true), this; },
}; };

// browser/window/context/viewport/document utils/stats
function _meta(ctx) {
    var extend = _extend(ctx),
        getSE = _util_mkScrollingElementGetter(ctx),
        scrollingElement,
        meta = {};

    if (getSE === true) { // no magic needed
        getSE = function() { return document.scrollingElement; };
        extend(meta, {
            seReread: getSE,
            se: getSE,
        });
    } else {
        extend(meta, {
            seReread: function() { return scrollingElement = getSE(); },
            se: function() {
                if (ctx.document.readyState != 'loading') {
                    meta.se = function() { return scrollingElement; };
                }
                return meta.seReread(); // warmup
            }
        });
    }

    // pageXYoffset to use with $rect(s)
    // see https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollY
    if ('pageXOffset' in ctx) extend(meta, {
        px: function() { return ctx.pageXOffset; },
        py: function() { return ctx.pageYOffset; },
    });
    else extend(meta, {
        px: function() { return meta.se().scrollLeft; },
        py: function() { return meta.se().scrollTop; },
    });

    return meta;
};

function $SAF(ctx) { // extend ctx with SAFDOM methods
    var extend = _extend(ctx),
        protoArr = _protoArr(ctx),
        document = ctx.document;

    extend(ctx.Node.prototype, _protoNode(ctx));
    extend(ctx.NodeList.prototype, protoArr);
    extend(ctx.HTMLCollection.prototype, protoArr);
    extend(ctx.Element.prototype, _protoElement(ctx));

    // wtf, ie? https://developer.mozilla.org/en-US/docs/Web/API/Event/target
    // so, EventTarget.prototype itself may not be available, therefore we need to extend individual things
    var protoET = _protoEventTarget(ctx);
    if (ctx.EventTarget) {
        extend(ctx.EventTarget.prototype, protoET);
    }
    else {
        // touch main things, leave others like XMLHttpRequest unmodified
        extend(ctx.self, protoET);
        extend(ctx.document, protoET);
        extend(ctx.Node.prototype, protoET);
    }

    return extend(ctx, {
    // Globals
    $id: document.getElementById.bind(document),
    $f: document.querySelectorAll.bind(document), // Find all => NONlive HTMLCollection [Chrome 1+, FF 3.5+, IE 8+, Opera 10+, Safari 3.2+]
    $s: document.querySelector.bind(document), // 1st matched element => Element or null [Chrome 1+, FF 3.5+, IE 8+, Opera 10+, Safari 3.2+]
    $t: document.getElementsByTagName.bind(document), // by Tag name => LIVE HTMLCollection [Chrome 1+, IE 6+]
    $c: document.getElementsByClassName.bind(document), // by Class name => LIVE HTMLCollection [IE 9+]
    $mk: document.createElement.bind(document),
    $mkT: document.createTextNode.bind(document),
    $mkF: document.createDocumentFragment.bind(document),
    //TODO $mk*, $mkEvent
    $meta: _meta(ctx),
    });
};

if (exports) module.exports = $SAF;
else $SAF(module);

})

// detect context type: if this is require()'d module -- provide export and do nothing more,
// otherwise (in browser) just extend window;
// seems tricky, but its simpler than it looks like:
.apply(null, (typeof module !== 'undefined' && module) ? [module, true] : [window, false]);
